Add Full support for roborock Zeo washing/drying machines

yangqian opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Breaking change
<!--
  If your PR contains a breaking change for existing users, it is important
  to tell them what breaks, how to make it work again and why we did this.
  This piece of text is published with the release notes, so it helps if you
  write it towards our users, not us.
  Note: Remove this section if this PR is NOT a breaking change.
-->


## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
The original roborock washing machine only exposes four sesnsors.
Here, this patch adds a bunch more. Not just sensors, but also buttons, selects, and switches.
| Configuration | Controls & Diagnostic |
| :---: | :---: |
| <img width="309" height="663" alt="Configuration" src="https://github.com/user-attachments/assets/e4d72acf-d4ff-4d30-946b-5460c6af21ab" /> | <img width="307" height="250" alt="Controls" src="https://github.com/user-attachments/assets/67bb488b-a531-489b-b6d0-97f7e06625f5" /><br><br><img width="309" height="387" alt="Diagnostic" src="https://github.com/user-attachments/assets/4ee7b6e0-01de-4d0b-84b9-eeacd6f71978" /> |

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [X] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: 
- Link to developer documentation pull request: 
- Link to frontend pull request: 

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.

  AI tools are welcome, but contributors are responsible for *fully*
  understanding the code before submitting a PR.
-->

- [X] I understand the code I am submitting and can explain how it works.
- [X] The code change is tested and works locally.
- [X] Local tests pass. **Your PR cannot be merged unless tests pass**
- [X] There is no commented out code in this PR.
- [X] I have followed the [development checklist][dev-checklist]
- [x] I have followed the [perfect PR recommendations][perfect-pr]
- [x] The code has been formatted using Ruff (`ruff format homeassistant tests`)
- [x] Tests have been added to verify that the new code works.
- [X] Any generated code has been carefully reviewed for correctness and compliance with project standards.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/development_checklist/
[manifest-docs]: https://developers.home-assistant.io/docs/creating_integration_manifest/
[quality-scale]: https://developers.home-assistant.io/docs/integration_quality_scale_index/
[docs-repository]: https://github.com/home-assistant/home-assistant.io
[perfect-pr]: https://developers.home-assistant.io/docs/review-process/#creating-the-perfect-pr

yangqian committed:
initial support for roborock washing machine

yangqian committed:
fix commented out code

yangqian committed:
Format A01 code: remove commented lines, fix trailing spaces, use lazy logging

yangqian committed:
ruff format

yangqian committed:
Refactor imports in Roborock components: remove unused imports and clean up code

yangqian committed:
Format button and switch code: improve readability by breaking long lines

allenporter reviewed (changes requested):
Thank you very much for this contribution.

NoRi2909 reviewed (commented):
Two strings need to be sentence-cased, the rest looks OK in this regard.

yangqian committed:
Address PR review comments for roborock washing machine integration

- Add missing translation keys for button_press_failed and select_option_failed exceptions
- Replace bare Exception with RoborockException in button, select, and switch entities
- Refactor select entity descriptions to use single enum_class field instead of options + option_values
- Fix select entity async_select_option to validate options and always call API with value
- Fix current_option to return string representation of current value
- Change sound_setting entity_category from DIAGNOSTIC to CONFIG
- Remove unused param field from RoborockButtonDescriptionA01
- Fix sentence case in strings.json (Wash and Dry -> Wash and dry, Rinse & Spin -> Rinse & spin)
- Remove 'zeo_' prefix from all translation keys (12 occurrences)
- Add value_fn support to sensor descriptions for boolean-to-string conversion
- Fix detergent_empty and softener_empty sensors to use 'empty'/'available' instead of True/False
- Add RoborockEnum import and type hints
- Add type: ignore comments for mypy attr-defined checks

allenporter commented on homeassistant/components/roborock/sensor.py:
```diff
@@ -261,13 +264,6 @@ def _dock_error_value_fn(state: DeviceState) -> str | None:
         device_class=SensorDeviceClass.ENUM,
         options=RoborockDyadStateCode.keys(),
     ),
-    RoborockSensorDescriptionA01(
```
why is this removed?

yangqian replied:
battery is not used in Zeo, but I guess it may show up in other devices, will revert.

allenporter commented on homeassistant/components/roborock/sensor.py:
```diff
@@ -287,7 +283,7 @@ def _dock_error_value_fn(state: DeviceState) -> str | None:
     ),
     RoborockSensorDescriptionA01(
-        key="error",
+        key="dyad_error",
```
i believe that changing this key will change unique ids which will break existing use cases. can this be reverted?

allenporter commented:
This PR makes also a bunch of other small changes to the existing code. Can you send them in logical separate PRs with description that give context/motivation?  They should be fast to review. But generally I'm left wondering what is intentional versus accidental AI changes.

allenporter commented on homeassistant/components/roborock/strings.json:
```diff
@@ -337,44 +511,6 @@
-          "water_level_error": "Water level error"
-        }
-      },
-      "zeo_state": {
```
I realize i told you to rename to not have "zeo" prefix but the existing key is using it. can you revert this so we don't lose translations?

allenporter replied:
I think my comment may have been a bad idea for some of the fields that have unique values to these specific devices, like the mode etc

yangqian commented on homeassistant/components/roborock/switch.py:
```diff
@@ -87,6 +110,17 @@ async def async_setup_entry(
+        )
+        for coordinator in config_entry.runtime_data.a01
+        for description in A01_SWITCH_DESCRIPTIONS
+        if description.data_protocol in coordinator.data
```
I tested my Zeo device, the switch is added correctly.

Sorry, I did not fully understand your comment. Do you mean to delete this async_add_entities for A01 switches and add "entity_registry_enabled_default=False" to A01_SWITCH_DESCRIPTIONS for "sound_setting"?
Could you make or suggest your change?

yangqian replied:
I only have one Zeo(A01) device, so I am not sure if sound_setting is universal or not. I'm now just commenting out the if statement.

yangqian committed:
revert changes on battery sensor and error descriptions for A01 devices

yangqian committed:
Remove unnecessary condition for A01 switch descriptions in setup entry

yangqian committed:
Rename A01 button descriptions to ZEO_BUTTON_DESCRIPTIONS and update setup logic for washing machine coordinator

joostlek reviewed (changes requested):
There's a merge conflict, can you take a look?

yangqian committed:
Format import statements for improved readability

yangqian committed:
Update homeassistant/components/roborock/strings.json

Co-authored-by: Norbert Rittel <norbert@rittel.de>

yangqian committed:
Update softener_type snapshot to match mock data (medium, not low)

yangqian committed:
Move detergent_type and softener_type from sensor to select

joostlek commented on homeassistant/components/roborock/button.py:
```diff
@@ -160,3 +201,34 @@ def __init__(
+            raise HomeAssistantError(
+                translation_domain=DOMAIN,
+                translation_key="button_press_failed",
+            ) from err
```
```suggestion
        except RoborockException as err:
            raise HomeAssistantError(
                translation_domain=DOMAIN,
                translation_key="button_press_failed",
            ) from err
        finally:
            await self.coordinator.async_request_refresh()
```

yangqian committed:
Update homeassistant/components/roborock/strings.json

Co-authored-by: Norbert Rittel <norbert@rittel.de>

yangqian committed:
Update homeassistant/components/roborock/strings.json

Co-authored-by: Norbert Rittel <norbert@rittel.de>

yangqian committed:
Merge branch 'dev' into roborock-cloud-push-for-A01-devices

yangqian committed:
Move Zeo detergent and softener empty to binary sensors

yangqian committed:
move async requet out of try block

yangqian committed:
Add ServiceValidationError for invalid select options in RoborockSelectEntityA01

yangqian committed:
Handle Roborock command failures with appropriate error messaging in RoborockSelectEntityA01

yangqian committed:
Add tests for A01 Zeo select entities including success and failure scenarios

yangqian committed:
Add detergent and softener states in translations

yangqian committed:
Add snapshot tests for roborock button and switch entities

yangqian committed:
Update binary sensor snapshots for washing machine entities

- Changed entity_id from binary_sensor.zeo_one_detergent_empty to binary_sensor.zeo_one_detergent
- Changed entity_id from binary_sensor.zeo_one_softener_empty to binary_sensor.zeo_one_softener
- Updated entity names from 'Detergent empty' to 'Detergent' and 'Softener empty' to 'Softener'
- Friendly names updated to match new entity naming

yangqian committed:
Add test coverage for A01 select entity None value handling

yangqian committed:
Update homeassistant/components/roborock/strings.json

Co-authored-by: Norbert Rittel <norbert@rittel.de>

yangqian committed:
Merge branch 'dev' into roborock-cloud-push-for-A01-devices

yangqian committed:
Merge branch 'dev' into roborock-cloud-push-for-A01-devices

joostlek reviewed (changes requested):
There's another merge conflict, can you take a look? 😅

yangqian committed:
Merge remote-tracking branch 'origin/dev' into roborock-cloud-push-for-A01-devices

# Conflicts:
#	homeassistant/components/roborock/select.py
#	homeassistant/components/roborock/sensor.py
#	homeassistant/components/roborock/strings.json

yangqian committed:
Remove unused RoborockDataUpdateCoordinatorB01 import from sensor.py

yangqian committed:
Filter A01 select and switch entities by supported data protocols

yangqian committed:
Remove unused Dyad Pro Sound setting snapshots from test_switch.ambr

yangqian committed:
Merge branch 'dev' into roborock-cloud-push-for-A01-devices

yangqian committed:
Remove unused error states and add aftercare status in Roborock strings

yangqian committed:
Merge branch 'dev' into roborock-cloud-push-for-A01-devices

joostlek committed:
Merge branch 'dev' into roborock-cloud-push-for-A01-devices

joostlek commented on homeassistant/components/roborock/strings.json:
```diff
@@ -198,6 +317,29 @@
+          "door_lock_error": "Door lock error",
+          "drain_error": "Drain error",
+          "drying_error": "Drying error",
+          "drying_error_e_12": "Drying error E12",
```
Can we add the error then? Maybe remove the E part and make sure the state also looks nice

yangqian replied:
OK. Added.

yangqian committed:
Add edge test for A01 switch handling unknown state when API omits protocol key

yangqian committed:
Update drying error messages for clarity in Roborock integration

yangqian committed:
Remove unused value_fn from RoborockSensorDescriptionA01

yangqian committed:
Merge branch 'dev' into roborock-cloud-push-for-A01-devices
